### PR TITLE
Mob-ifys Turtles

### PR DESCRIPTION
**Pull request by JORJ949**
## About the PR 
Makes turtles (like sylvesters) mobs
Also allows any turtle to wear any beret, particular turtles only allow certain jobs to remove their hats.

## Why's this needed? 
Less obj/critter is more good. 

First time messing with critter code so there's probably a ton of issues I missed.

## Changelog 

```changelog
(u)JORJ949
(+)Turtles are now mobs, you can put them in pet carriers!
```

### CODE REVIEW

**Comment by Glamurio** [comment]
Huge W for the turtle enjoyers

File: code/mob/living/critter/small_animal.dm [hunk 1]
```diff
@@ -338,6 +338,383 @@ proc/filter_carrier_pets(var/type)
...
+		var/list/hearers_list = hearers(range, src)
+		for (var/mob/living/M in hearers_list)
+			if (istype(M, /mob/living/carbon/human))
+				var/mob/living/carbon/human/human = M
+				var/clown_tally = human.clown_tally()
+				if (isalive(human) && (clown_tally>=2 || human.traitHolder.hasTrait("training_clown"))) //We only hate clowns
+					. += human
+		if (length(.))
+			if (!ON_COOLDOWN(src,"clown_charge_alert", 2 MINUTES))
+				src.visible_message(SPAN_ALERT("<b>[src]</b> notices a Clown and starts charging!"))
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:536]: Shouldn't chasing clowns only be on Sylvester?

File: code/mob/living/critter/small_animal.dm [hunk 2]
```diff
@@ -338,6 +338,383 @@ proc/filter_carrier_pets(var/type)
...
+			return 1
+		return 0
+
+//The HoS's pet turtle. He can wear the beret!
+/mob/living/critter/small_animal/turtle/sylvester
+	name = "Sylvester"
+	desc = "This turtle looks both cute and indimidating. It's a tough line to walk, but he does it effortlessly."
+	health_brute = 50
+	health_burn = 50
+	gender = MALE
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:678]:
```suggestion
	gender = MALE
	player_can_spawn_with_pet = FALSE
```
no cloning the HoS's pet!

File: code/mob/living/critter/small_animal.dm [hunk 3]
```diff
@@ -338,6 +338,383 @@ proc/filter_carrier_pets(var/type)
...
+	proc/enter_shell()
+		if (src.shell_count) return 0
+		src.shell_count = 10
+		src.ai.disable()
+
+		APPLY_ATOM_PROPERTY(src, PROP_MOB_EXPLOPROT, "turtle_shell", 80)
+		APPLY_ATOM_PROPERTY(src, PROP_MOB_MELEEPROT_HEAD, "turtle_shell", 80)
+		APPLY_ATOM_PROPERTY(src, PROP_MOB_MELEEPROT_BODY, "turtle_shell", 80)
+
+		icon_state = "[base_icon_state]-shell"
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:609]: This doesn't seem to be working?

File: code/mob/living/critter/small_animal.dm [hunk 4]
```diff
@@ -338,6 +338,383 @@ proc/filter_carrier_pets(var/type)
...
+		icon_state = base_icon_state
+		if (costume_name)
+			src.UpdateOverlays(costume_alive, "costume")
+		density = FALSE
+		UpdateIcon()
+		src.visible_message(SPAN_NOTICE("<b>[src]</b> comes out of [his_or_her(src)] shell!"))
+		return 1
+
+
+	proc/give_beret(var/obj/hat, var/mob/user)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:635]: Sylvester seems to revive with his beret sprite, which is I think what this is for, should probably check that on revive somehow?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/mob/living/critter/small_animal.dm (hunk 4 above)
Comment (JORJ949) [reply code/mob/living/critter/small_animal.dm:622]: Im not sure what gets called where when a mob is revived (e.g. where it sets the icon back to the initial one) but it should definitely call UpdateIcon()

File: code/mob/living/critter/small_animal.dm [hunk 5]
```diff
@@ -522,19 +524,6 @@ proc/filter_carrier_pets(var/type)
...
-		for (var/mob/living/M in hearers_list)
-			if (istype(M, /mob/living/carbon/human))
-				var/mob/living/carbon/human/human = M
-				var/clown_tally = human.clown_tally()
-				if (isalive(human) && (clown_tally>=2 || human.traitHolder.hasTrait("training_clown"))) //We only hate clowns
-					. += human
-		if (length(.))
-			if (!ON_COOLDOWN(src,"clown_charge_alert", 2 MINUTES))
-				src.visible_message(SPAN_ALERT("<b>[src]</b> notices a Clown and starts charging!"))
-
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm]: Just removing this has made normal turtles aggressive to everyone, you'll need to change their behaviour to be retaliate-only as well.

File: code/mob/living/critter/ai/turtle.dm [hunk 6]
```diff
@@ -0,0 +1,9 @@
+/datum/aiHolder/turtle
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/turtle, list(src))
+
+/datum/aiTask/prioritizer/critter/turtle/New()
+	..()
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/critter/attack, list(holder, src))
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/ai/turtle.dm:9]: Actually shouldn't this just use the standard `aggressive` AI holder?

File: code/mob/living/critter/small_animal.dm [hunk 7]
```diff
@@ -338,6 +339,385 @@ proc/filter_carrier_pets(var/type)
...
+	bullet_act(var/obj/projectile/P)
+		switch(P.proj_data.damage_type)
+			if(D_KINETIC,D_PIERCING,D_SLASHING)
+				if (prob(70))
+					src.enter_shell()
+		..()
+
+	attack_hand(mob/user)
+		if (isdead(src))
+			take_beret(user)
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:447]: This still doesn't actually work, you'll need to either allow the beret to be taken while the turtle is in their shell or make them exit their shell when they die.

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

File: code/mob/living/critter/small_animal.dm (hunk 7 above)
Comment (JORJ949) [reply code/mob/living/critter/small_animal.dm:447]: This was copypasted from old turtle code, im gonna remove it as you can still clickdrag

File: code/mob/living/critter/small_animal.dm [hunk 8]
```diff
@@ -513,17 +508,9 @@ proc/filter_carrier_pets(var/type)
...
-			return 1
-		if (src.ai?.enabled && ishuman(user))
-			var/mob/living/carbon/human/human = user
-			var/clown_tally = human.clown_tally()
-			if (clown_tally>=2 || human.traitHolder.hasTrait("training_clown"))
-				src.ai.priority_tasks += src.ai.get_instance(/datum/aiTask/sequence/goalbased/critter/attack, list(src, src.ai.default_task))
-				src.ai.interrupt()
-				src.visible_message(SPAN_NOTICE("[src] notices a Clown and knocks [human] over!"))
-				human.setStatus("resting", duration = INFINITE_STATUS)
+	updatehealth()
```
Comment (TobleroneSwordfish) [inline code/mob/living/critter/small_animal.dm:511]: Don't put this here, calling `UpdateIcon` every time something takes damage is wasteful. Why not just put it in `death` and `full_heal`?

**Review by TobleroneSwordfish — CHANGES_REQUESTED** [review]

**Pull request merged**